### PR TITLE
Push/Pull cancellability UTs

### PR DESCRIPTION
**Pull request by shrishrirang**
- Added UTs to test cancellability of push and pull operations at various execution points
- Removed UT testPullResultingInErrorAllowsDependentOperationsToComplete as completion of push/pull operation is covered by the newly added UTs

### CODE REVIEW

File: sdk/iOS/test/MSSyncTableTests.m [hunk 1]
```diff
@@ -1878,54 +2218,6 @@ -(void) testIncrementalPullWithOrderByFails
     XCTAssertEqual(1, filter.actualRequests.count);
 }
 
-- (void)testPullResultingInErrorAllowsDependentOperationsToComplete
```
Comment (phvannor) [inline sdk/iOS/test/MSSyncTableTests.m]: Why is this test removed?

File: sdk/iOS/test/MSSyncTableTests.m [hunk 2]
```diff
@@ -1082,9 +1083,348 @@ -(void) testPushNetworkTimeoutWithErroredOps
     XCTAssertEqual(filteredClient.syncContext.pendingOperationsCount, 3);
 }
 
+// Tests cancellation of the push operation at various stages of a push workflow.
+// - Verifies that a push following a cancelled push operation works as expected
+// - Verifies that cancelling a push operation finishes the push NSOperation regardless of the cancellation point
+// - Verifies that the expected number of items are pushed to the server
+-(void) testPushCancellability
```
Comment (phvannor) [inline sdk/iOS/test/MSSyncTableTests.m]: I don't like this style of test for unit tests, it doesn't play nice (in my opinion) with the FW involved.

File: sdk/iOS/test/MSSyncTableTests.m [hunk 3]
```diff
@@ -1082,9 +1083,348 @@ -(void) testPushNetworkTimeoutWithErroredOps
...
+// - Returns the number of items pushed to the server.
+-(int) performPushWithCancellationPoint:(int) requestedCancellationPoint serverResponseCode:(NSNumber *)serverResponseCode
+{
+    // Initialization
+    MSMultiRequestTestFilter *filter = [MSMultiRequestTestFilter testFilterWithStatusCodes:@[serverResponseCode] data:@[@"{\"id\": \"id1\", \"text\":\"server\"}"] appendEmptyRequest:YES];
+    MSClient *filteredClient = [client clientWithFilter:filter];
+    MSSyncTable *todoTable = [filteredClient syncTableWithName:TodoTableNoVersion];
+
+    // Start with a clean slate
+    [[todoTable forcePurgeWithCompletion:^(NSError *error) {
```
Comment (phvannor) [inline sdk/iOS/test/MSSyncTableTests.m]: Tests use an in memory store, why did you need to add a force purge here?

File: sdk/iOS/test/MSSyncTableTests.m [hunk 4]
```diff
@@ -1082,9 +1083,348 @@ -(void) testPushNetworkTimeoutWithErroredOps
...
+        XCTAssertNil(error, @"error should have been nil.");
+    }] waitUntilFinished];
+
+    __block NSOperation *pushOperation;
+
+    // Define filters
+    ((MSTestFilter *)filter.testFilters[0]).onInspectRequest = ^(NSURLRequest *request) {
+
+        // Cancellation point 0
+        if (requestedCancellationPoint == 0)
```
Comment (phvannor) [inline sdk/iOS/test/MSSyncTableTests.m]: nit: style, always use { }

File: sdk/iOS/test/MSSyncTableTests.m [hunk 5]
```diff
@@ -1082,9 +1083,348 @@ -(void) testPushNetworkTimeoutWithErroredOps
...
+    }];
+    
+    [self waitForExpectationsWithTimeout:10 handler:nil];
+    
+    // Perform a push
+    pushOperation = [filteredClient.syncContext pushWithCompletion:^(NSError *error) {
+        
+        // Cancellation point 2
+        if (requestedCancellationPoint == 2)
+            [pushOperation cancel];
```
Comment (phvannor) [inline sdk/iOS/test/MSSyncTableTests.m]: this is cancelling after the op has finished?  Is that where you intended to do this?  Or did you mean to get one for the delegate version of push complete?

File: sdk/iOS/test/MSSyncTableTests.m [hunk 6]
```diff
@@ -1082,9 +1083,348 @@ -(void) testPushNetworkTimeoutWithErroredOps
...
+
+// Performs push and cancels it at the specified cancellation point.
+// - Verifies that a push operation following a cancelled push operation works as expected
+// - Verifies that cancelling a push finishes the push NSOperation regardless of the cancellation point
+// - Verifies that the expected number of items are pushed to the server
+// - Returns the number of items pushed to the server.
+-(int) performPushWithCancellationPoint:(int) requestedCancellationPoint serverResponseCode:(NSNumber *)serverResponseCode
+{
+    // Initialization
+    MSMultiRequestTestFilter *filter = [MSMultiRequestTestFilter testFilterWithStatusCodes:@[serverResponseCode] data:@[@"{\"id\": \"id1\", \"text\":\"server\"}"] appendEmptyRequest:YES];
```
Comment (phvannor) [inline sdk/iOS/test/MSSyncTableTests.m]: Why do you need to use a multirequest filter?  I only counted the one push.

File: sdk/iOS/test/MSSyncTableTests.m [hunk 7]
```diff
@@ -1082,9 +1083,348 @@ -(void) testPushNetworkTimeoutWithErroredOps
...
+{
+    // Define a filter that returns an appropriate server response for the requested item
+    MSTestFilter *filter = [MSTestFilter new];
+    MSClient *filteredClient = [client clientWithFilter:filter];
+
+    filter.ignoreNextFilter = YES;
+    filter.onInspectResponseData = ^(NSURLRequest *request, NSData *data) {
+        NSError *error = nil;
+        
+        NSDictionary *item = [filteredClient.serializer itemFromData:request.HTTPBody withOriginalItem:nil ensureDictionary:YES orError:&error];
```
Comment (phvannor) [inline sdk/iOS/test/MSSyncTableTests.m]: This isn't a safe assumption in a generic push method, deletes don't send an item in them.  Ideally push should have a couple short unit tests verifying shape & body, not mixed in with cancel tests.

File: sdk/iOS/test/MSSyncTableTests.m [hunk 8]
```diff
@@ -1082,9 +1083,348 @@ -(void) testPushNetworkTimeoutWithErroredOps
...
+
+    return synchronizedItemCount;
+}
+
+// Performs a push operation and returns the count of all client items pushed so far
+-(int) performVanillaPush
+{
+    // Define a filter that returns an appropriate server response for the requested item
+    MSTestFilter *filter = [MSTestFilter new];
+    MSClient *filteredClient = [client clientWithFilter:filter];
```
Comment (phvannor) [inline sdk/iOS/test/MSSyncTableTests.m]: This ends up trumping the filter setup by the actual test case, I'm a little lost as to the intent here.

File: sdk/iOS/test/MSSyncTableTests.m [hunk 9]
```diff
@@ -1082,9 +1083,348 @@ -(void) testPushNetworkTimeoutWithErroredOps
...
+    }];
+    [self waitForExpectationsWithTimeout:10 handler:nil];
+    
+    return [self synchronizedItemCount];
+}
+
+// Returns the number of synchronized items, i.e. items that are present on the server as well as the client.
+// An item is synchronized if it was successfully pushed to the server from the client OR
+// if it was successfully pulled from the server to the client.
+-(int) synchronizedItemCount
```
Comment (phvannor) [inline sdk/iOS/test/MSSyncTableTests.m]: Why do you need this?  The test offline store, already tracks a number of simple metrics.  What are we gaining by adding a read here as well?

File: sdk/iOS/test/MSSyncTableTests.m [hunk 10]
```diff
@@ -1082,9 +1083,348 @@ -(void) testPushNetworkTimeoutWithErroredOps
...
+    
+    int synchronizedItemCount = [self synchronizedItemCount];
+    
+    XCTAssertEqual([self performVanillaPull], 3, "Expected all server items to have been pulled to the client by now");
+    
+    return synchronizedItemCount;
+}
+
+// Performs a pull operation and returns the count of all the server items pulled so far
+-(int) performVanillaPull
```
Comment (phvannor) [inline sdk/iOS/test/MSSyncTableTests.m]: We really care more about the incremental pulls than the defaults.

File: sdk/iOS/test/MSSyncTableTests.m [hunk 11]
```diff
@@ -1082,9 +1083,348 @@ -(void) testPushNetworkTimeoutWithErroredOps
...
+    MSQuery *query = [[MSQuery alloc] initWithSyncTable:todoTable];
+
+    __block bool isFirstPullRequest = true;
+    filter.ignoreNextFilter = YES;
+    filter.onInspectResponseData = ^(NSURLRequest *request, NSData *data) {
+
+        NSDictionary *item = [filteredClient.serializer itemFromData:request.HTTPBody withOriginalItem:nil ensureDictionary:YES orError:nil];
+        
+        NSString *responseString;
+        if (item != nil) // This means it is a push request
```
Comment (phvannor) [inline sdk/iOS/test/MSSyncTableTests.m]: This is why the multirequest filter exists, to let the filter handle returning responses/etc.

File: sdk/iOS/test/MSSyncTableTests.m (hunk 1 above)
Comment (shrishrirang) [inline sdk/iOS/test/MSSyncTableTests.m]: This test verifies that the pull NSOperation is in finished state if the push fails. There is just one of many cases (error at different points in the workflow and not just during push, no error case, etc) in which we need to test for successful completion of the NSOperation. All this is being already covered as part of the 2 push and pull UTs added.

File: sdk/iOS/test/MSSyncTableTests.m (hunk 2 above)
Comment (shrishrirang) [inline sdk/iOS/test/MSSyncTableTests.m]: I agree in general it is non-ideal to couple multiple tests. However, in this case, we are essentially just testing cancel. The 3 verification bullets listed are all attributes of a clean cancel and not really different test cases.

Comment (phvannor) [inline sdk/iOS/test/MSSyncTableTests.m]: I'd much rather see each case broken out, and named in a readable manner.  I'd prefer no shared code involving asserts, but at a minimum it should be immediately obv what one failed & simple to replay a single case.

Comment (shrishrirang) [inline sdk/iOS/test/MSSyncTableTests.m]: OK, I am planning of breaking down the push test into the following (and similar for pull). This way there are 2 tests:
- testpush_200: a successful push that gets cancelled at various possible points during its execution
- testpush_500: an unsuccessful push that gets cancelled at various possible points during its execution

Shared code will continue to have assertions that should hold true irrespective of the cancellation point and error codes.

-(void) testPushCancellability_200
{
    XCTAssertEqual([self performPushWithCancellationPoint:0 serverResponseCode:@200], 0, "Push cancellation failed");
    XCTAssertEqual([self performPushWithCancellationPoint:1 serverResponseCode:@200], 0, "Push cancellation failed");
    XCTAssertEqual([self performPushWithCancellationPoint:2 serverResponseCode:@200], 1, "Push cancellation failed");
    XCTAssertEqual([self performPushWithCancellationPoint:3 serverResponseCode:@200], 0, "Push cancellation failed");
}

-(void) testPushCancellability_500
{
    XCTAssertEqual([self performPushWithCancellationPoint:0 serverResponseCode:@500], 0, "Push cancellation failed");
    XCTAssertEqual([self performPushWithCancellationPoint:1 serverResponseCode:@500], 0, "Push cancellation failed");
    XCTAssertEqual([self performPushWithCancellationPoint:2 serverResponseCode:@500], 0, "Push cancellation failed");
    XCTAssertEqual([self performPushWithCancellationPoint:3 serverResponseCode:@500], 0, "Push cancellation failed");
}

File: sdk/iOS/test/MSSyncTableTests.m (hunk 3 above)
Comment (shrishrirang) [inline sdk/iOS/test/MSSyncTableTests.m]: The in-memory store will be initialized before testPushCancellability is called, but we are calling performPushWithCancellationPoint multiple times from within it.

File: sdk/iOS/test/MSSyncTableTests.m (hunk 2 above)
Comment (phvannor) [inline sdk/iOS/test/MSSyncTableTests.m]: Its still hard from that to say why it failed, I'd prefer at the least to have the assert describe the failure reason.  I'd actually just like each assert to be its own test function really for better interplay with the bots & reporting.

File: sdk/iOS/test/MSSyncTableTests.m (hunk 5 above)
Comment (shrishrirang) [inline sdk/iOS/test/MSSyncTableTests.m]: The completion callback is called asynchronously before the operation is completed, so the operation can be in progress when we cancel.

File: sdk/iOS/test/MSSyncTableTests.m (hunk 6 above)
Comment (shrishrirang) [inline sdk/iOS/test/MSSyncTableTests.m]: In an earlier iteration two items were inserted, not anymore. Fixed it.

File: sdk/iOS/test/MSSyncTableTests.m (hunk 7 above)
Comment (shrishrirang) [inline sdk/iOS/test/MSSyncTableTests.m]: This is not a generic push. The idea is to check that a cancelled push does not put the internal data structures in a weird state such that no future push operations go through. This method simply performs a very basic (but not generic) push and ensures that it completes fine. If the method name appears confusing, I can change that.

File: sdk/iOS/test/MSSyncTableTests.m (hunk 8 above)
Comment (shrishrirang) [inline sdk/iOS/test/MSSyncTableTests.m]: Yes, that's intentional. This push is performed without cancellation after a prior push operation was cancelled. The idea is to see if this completes successfully to verify that the cancelled push operation did not mess up internal data structures such that no future push operations go through.

File: sdk/iOS/test/MSSyncTableTests.m (hunk 9 above)
Comment (shrishrirang) [inline sdk/iOS/test/MSSyncTableTests.m]: This method is used to count the number of items that have been pushed (or pulled) at any point of time (as part of all the past push or pull operations).

File: sdk/iOS/test/MSSyncTableTests.m (hunk 10 above)
Comment (shrishrirang) [inline sdk/iOS/test/MSSyncTableTests.m]: Not sure I understand your comment. We are not testing incremental pull here. We are just making sure future pull operations continue to work fine after cancelling a pull. This method just performs a very basic pull.

File: sdk/iOS/test/MSSyncTableTests.m (hunk 11 above)
Comment (shrishrirang) [inline sdk/iOS/test/MSSyncTableTests.m]: This method is called more than once and the responses expected are a function of when cancel was performed. We don't want a statically configured filter.

File: sdk/iOS/test/MSSyncTableTests.m (hunk 7 above)
Comment (phvannor) [inline sdk/iOS/test/MSSyncTableTests.m]: Yes, this method is confusing to me.  Its defintely not a vanilla push then, and its assuming alot of internal state making it only work for specific test cases that is unclear why I would use it.

File: sdk/iOS/test/MSSyncTableTests.m (hunk 8 above)
Comment (phvannor) [inline sdk/iOS/test/MSSyncTableTests.m]: Its an odd mechanic for a helper to override the parent's filter, esp given push is really dependent on the parent's actions up to now.

File: sdk/iOS/test/MSSyncTableTests.m (hunk 9 above)
Comment (phvannor) [inline sdk/iOS/test/MSSyncTableTests.m]: The helper gives you base stats on upserted items, is this count going to be different?

File: sdk/iOS/test/MSSyncTableTests.m [hunk 12]
```diff
@@ -1082,9 +1083,348 @@ -(void) testPushNetworkTimeoutWithErroredOps
...
+    __block int numSynchronizedItems = 0;
+
+    XCTestExpectation *readExpectation = [self expectationWithDescription:@"read expectation"];
+
+    // Read from the table
+    [client.syncContext readWithQuery:query completion:^(MSQueryResult *result, NSError *error) {
+        
+        // Count the number of synchronized items
+        for (NSDictionary* item in result.items) {
+            if ([item[@"text"] isEqualToString:@"server"]) {
```
Comment (phvannor) [inline sdk/iOS/test/MSSyncTableTests.m]: This should just be made part of the query, no reason to loop over results when you could let the store do the filtering.

File: sdk/iOS/test/MSSyncTableTests.m [hunk 13]
```diff
@@ -1082,9 +1083,348 @@ -(void) testPushNetworkTimeoutWithErroredOps
...
+        [readExpectation fulfill];
+    }];
+    
+    [self waitForExpectationsWithTimeout:10 handler:nil];
+    
+    return numSynchronizedItems;
+}
+
+// Verify that the specified NSOperation has finished
+-(void) verifyFinished:(NSOperation *)operation
```
Comment (phvannor) [inline sdk/iOS/test/MSSyncTableTests.m]: Cant this just use a completionblock on the NSOperation itself to verify its complete?

File: sdk/iOS/test/MSSyncTableTests.m (hunk 10 above)
Comment (phvannor) [inline sdk/iOS/test/MSSyncTableTests.m]: Just that we don't expect ppl to use the basic pull, only inc pulls in the wild.  So unit test coverage can focus on the inc pull codepaths if we don't cover both sets in UTs

File: sdk/iOS/test/MSSyncTableTests.m (hunk 11 above)
Comment (phvannor) [inline sdk/iOS/test/MSSyncTableTests.m]: Which is why I don't understand overriding the filter that is setup by the base unit test, if the responses adjust dynamically anyways its unclear what we are testing.

File: sdk/iOS/test/MSSyncTableTests.m (hunk 2 above)
Comment (shrishrirang) [inline sdk/iOS/test/MSSyncTableTests.m]: Ok, I'll make the assert message more detailed.
I'm avoiding splitting the 4 calls in into methods of their own because the only difference in them would be the cancellation points and they are kinda arbitrary points in the execution workflow. (Same for pull).

File: sdk/iOS/test/MSSyncTableTests.m (hunk 12 above)
Comment (shrishrirang) [inline sdk/iOS/test/MSSyncTableTests.m]: Done.

File: sdk/iOS/test/MSSyncTableTests.m (hunk 13 above)
Comment (shrishrirang) [inline sdk/iOS/test/MSSyncTableTests.m]: The NSOperation returned by the push/pull methods could potentially be already finished. We may not get a chance to add the completion block to the NSOp before it is finished.

File: sdk/iOS/test/MSSyncTableTests.m (hunk 9 above)
Comment (shrishrirang) [inline sdk/iOS/test/MSSyncTableTests.m]: Do you mean the upsertedItems, upsertedCalls, etc metrics? They track the upserted items in all tables (the table we are inserting in, operations table, errors table) not just the table we are inserting in. We just care about the final state of the user table.

File: sdk/iOS/test/MSSyncTableTests.m (hunk 7 above)
Comment (shrishrirang) [inline sdk/iOS/test/MSSyncTableTests.m]: Updated the method name.

File: sdk/iOS/test/MSSyncTableTests.m (hunk 8 above)
Comment (shrishrirang) [inline sdk/iOS/test/MSSyncTableTests.m]: The caller will now decide which client is to be used. The method now accepts the client as a param.

File: sdk/iOS/test/MSSyncTableTests.m (hunk 10 above)
Comment (shrishrirang) [inline sdk/iOS/test/MSSyncTableTests.m]: The UT is verifying that the expected items are inserted in the local store after the cancelled push is completed. I think it should suffice to just rely on the fact that the cancelled pull leaves the data store in the correct state.

File: sdk/iOS/test/MSSyncTableTests.m (hunk 11 above)
Comment (shrishrirang) [inline sdk/iOS/test/MSSyncTableTests.m]: There is a difference in what the first filter does and what the filter in performVanillaPush (method renamed now) does.
The first filter expects a push request for the inserted item followed by a pull request and is statically configured to respond accordingly.
The second filter may get variable number of push requests depending on when the earlier pull was cancelled. This filter looks at the requests and responds accordingly dynamically. This is the reason for overriding the first filter.

File: sdk/iOS/test/MSSyncTableTests.m (hunk 13 above)
Comment (phvannor) [inline sdk/iOS/test/MSSyncTableTests.m]: So can't we just call [operation waitUntilFinished]?

File: sdk/iOS/test/MSSyncTableTests.m [hunk 14]
```diff
@@ -1082,9 +1083,380 @@ -(void) testPushNetworkTimeoutWithErroredOps
...
 
 #pragma mark Pull Tests
 
+// Tests cancellation of the pull operation at various stages of the pull workflow
+// - Verifies that a pull operation following a cancelled pull operation works as expected
+// - Verifies that cancelling a pull operation finishes the pull NSOperation regardless of the cancellation point
+// - Verifies that the expected number of items are pulled from the server
+-(void) testPullCancellability_successfulpush_successfulpull
+{
+    XCTAssertEqual([self performPullWithCancellationPoint:0 serverPushResponseCode:@200 serverPullResponseCode:@200], 1, "Pull cancellation failed at cancellation point 0");
```
Comment (phvannor) [inline sdk/iOS/test/MSSyncTableTests.m]: I would like better comments on each point, so its clear from reading the code what each "point" is testing.  Its unclear now to just have code say == 0, 1, etc and later on we dont know why these points where picked.

File: sdk/iOS/test/MSSyncTableTests.m [hunk 15]
```diff
@@ -1082,9 +1083,380 @@ -(void) testPushNetworkTimeoutWithErroredOps
...
+    [self waitForExpectationsWithTimeout:10 handler:nil];
+}
 
 #pragma mark Pull Tests
 
+// Tests cancellation of the pull operation at various stages of the pull workflow
+// - Verifies that a pull operation following a cancelled pull operation works as expected
+// - Verifies that cancelling a pull operation finishes the pull NSOperation regardless of the cancellation point
+// - Verifies that the expected number of items are pulled from the server
+-(void) testPullCancellability_successfulpush_successfulpull
```
Comment (phvannor) [inline sdk/iOS/test/MSSyncTableTests.m]: I would still prefer each cancel point be its own test method.  Its unclear what is gained by combining them like so. Now you are dealing with previous test state/etc that would just be avoided by breaking them all up.

File: sdk/iOS/test/MSSyncTableTests.m (hunk 13 above)
Comment (shrishrirang) [inline sdk/iOS/test/MSSyncTableTests.m]: we want to verify that it indeed finishes. waituntilfinish will block forever if the nsop hasn't finished.

File: sdk/iOS/test/MSSyncTableTests.m (hunk 14 above)
Comment (shrishrirang) [inline sdk/iOS/test/MSSyncTableTests.m]: Agreed. Done.

File: sdk/iOS/test/MSSyncTableTests.m (hunk 15 above)
Comment (shrishrirang) [inline sdk/iOS/test/MSSyncTableTests.m]: Done

**Comment by shrishrirang** [comment]
Ping..
